app/questionnaire/questionnaire_manager.py

```diff
@@ -178,134 +44,76 @@ def validate(self, post_data):
-            else:
-                # bug fix for using back button which then fails validation
-                self.go_to_node(self.get_current_location())
 
-            # now return the location
-            current_location = self.get_current_location()
-            logger.debug("Returning location %s", current_location)
-            return is_valid
-        else:
-            raise InvalidLocationException()
+        is_valid = self.validate(location, post_data)
+        # run the validator to update the validation_store
+        if is_valid:
+
+            # Store answers in QuestionnaireStore
+            questionnaire_store = get_questionnaire_store(current_user.user_id, current_user.user_ik)
+
+            for answer in self.get_answers(location):
+                questionnaire_store.answers[answer.id] = answer.value
+
+            if location not in questionnaire_store.completed_blocks:
+                questionnaire_store.completed_blocks.append(location)
+
+            questionnaire_store.save()
+
+        return is_valid
 
     def get_rendering_context(self, location, is_valid=True):
-        node = self._current
-        state_items = []
+
         if is_valid:
             if location == 'summary':
                 return self.get_summary_rendering_context()
             else:
-                self.build_state(node, node.answers)
+                # apply page answers?
+                self.build_state(location, get_answers(current_user))
 
         if self.state:
-            self._plumbing_preprocessing(self.state)
+            self._plumbing_preprocessing()
             self._conditional_display(self.state)
 
-        if self.state not in state_items:
-            state_items.append(self.state)
-
         # look up the preprocessor and then build the view data
-        return build_questionnaire_model(self._json, state_items)
+        return build_questionnaire_model(self._json, self.state)
 
     def get_summary_rendering_context(self):
         schema_template_context = get_schema_template_context(self, self._schema)
         rendered_questionnaire_schema_json = render_template_string(json.dumps(self._json), **schema_template_context)
+
         # look up the preprocessor and then build the view data
         return build_summary_model(json.loads(rendered_questionnaire_schema_json))
 
-    def build_state(self, node, answers):
-        # Build the state from the linked list and the answers
+    def build_state(self, item_id, answers):
+        # Build the state from the answers
         self.state = None
-        if self._schema.item_exists(node.item_id):
-            schema_item = self._schema.get_item_by_id(node.item_id)
+        if self._schema.item_exists(item_id):
+            schema_item = self._schema.get_item_by_id(item_id)
             self.state = schema_item.construct_state()
             self.state.update_state(answers)
 
-    def _plumbing_preprocessing(self, state):
+    def get_answers(self, item_id):
```

Comment:
I think this method name could be confused with the global `get_answers`, `get_state_answers` or `get_answers_from_state` would be better.